[receiver/receiver_creator] Add support for enabling receivers/scrapers from K8s hints

ChrsMark opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
## Description

~~This PR implements the metrics' part suggested at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34427#issue-2451363574. The current implementation only supports hints coming from Pods (wip with Logs included-> https://github.com/open-telemetry/opentelemetry-collector-contrib/compare/main...ChrsMark:opentelemetry-collector-contrib:logs_hints?expand=1)~~

This PR implements what it was decided at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34427#issuecomment-2436140051.

See the README docs for the description of this feature: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-4127365c4062a7510fb7fede0fa239e9232549732898303d94c12fef0433d39d

### TODOs:

- [x] turn-on/off the feature using a setting
- [x] add docs
- [x] add tests



<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
## Link to tracking issue
Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34427

<!--Describe what testing was performed and which tests were added.-->
## Testing
Added unit-tests.

<!--Describe the documentation added.-->
## Documentation
Added


## How to test this locally


1. Using the follow values to install the Collector's Helm chart:
```yaml
mode: daemonset

image:
  repository: otelcontribcol-dev
  tag: "latest"
  pullPolicy: IfNotPresent

command:
  name: otelcontribcol

clusterRole:
  create: true
  rules:
   - apiGroups:
     - ''
     resources:
     - 'pods'
     - 'nodes'
     verbs:
     - 'get'
     - 'list'
     - 'watch'
   - apiGroups: [ "" ]
     resources: [ "nodes/proxy"]
     verbs: [ "get" ]
   - apiGroups:
       - ""
     resources:
       - nodes/stats
     verbs:
       - get
   - nonResourceURLs:
       - "/metrics"
     verbs:
       - get

config:
  extensions:
    k8s_observer:
      auth_type: serviceAccount
      node: ${env:K8S_NODE_NAME}
      observe_nodes: true
  exporters:
    debug:
      verbosity: basic
  receivers:
    receiver_creator/metrics:
      watch_observers: [ k8s_observer ]
      discovery:
        enabled: true
      receivers:

  service:
    extensions: [health_check, k8s_observer]
    telemetry:
      logs:
        level: DEBUG
    pipelines:
      metrics:
        receivers: [ receiver_creator/metrics ]
        processors: [ batch ]
        exporters: [ debug ]
```

2. Use the following target Pod which consists of 2 containers:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: nginx-conf
data:
  nginx.conf: |
    user  nginx;
    worker_processes  1;
    error_log  /dev/stderr warn;
    pid        /var/run/nginx.pid;
    events {
      worker_connections  1024;
    }
    http {
      include       /etc/nginx/mime.types;
      default_type  application/octet-stream;
    
      log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                        '$status $body_bytes_sent "$http_referer" '
                        '"$http_user_agent" "$http_x_forwarded_for"';
      access_log  /dev/stdout main;
      server {
          listen 80;
          server_name localhost;
    
          location /nginx_status {
              stub_status on;
          }
      }
      include /etc/nginx/conf.d/*;
    }
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: redis-deployment
  labels:
    app: redis
spec:
  replicas: 1
  selector:
    matchLabels:
      app: redis
  template:
    metadata:
      labels:
        app: redis
      annotations:
        # redis container port hints
        io.opentelemetry.discovery.metrics.6379/enabled: "true"
        io.opentelemetry.discovery.metrics.6379/scraper: redis
        io.opentelemetry.discovery.metrics.6379/config: |
          collection_interval: "20s"
          timeout: "10s"

        # nginx container port hints
        io.opentelemetry.discovery.metrics.80/enabled: "true"
        io.opentelemetry.discovery.metrics.80/scraper: nginx
        io.opentelemetry.discovery.metrics.80/config: |
          endpoint: "http://`endpoint`/nginx_status"
          collection_interval: "30s"
          timeout: "20s"
    spec:
      volumes:
      - name: nginx-conf
        configMap:
          name: nginx-conf
          items:
            - key: nginx.conf
              path: nginx.conf
      containers:
        - name: webserver
          image: nginx:latest
          ports:
            - containerPort: 80
              name: webserver
          volumeMounts:
            - mountPath: /etc/nginx/nginx.conf
              readOnly: true
              subPath: nginx.conf
              name: nginx-conf
        - image: redis
          imagePullPolicy: IfNotPresent
          name: redis
          ports:
            - name: redis
              containerPort: 6379
              protocol: TCP
```

3. After the target pod is deployed check the Collector logs and verify that 2 receivers are started properly and metrics are collected from both target containers:
```console
2024-10-04T13:53:02.550Z	info	receivercreator@v0.110.0/observerhandler.go:115	starting receiver	{"kind": "receiver", "name": "receiver_creator/metrics", "data_type": "metrics", "name": "nginx/e6d880c9-13a0-45e7-a3a1-27fa021802b4_80", "endpoint": "10.244.0.52:80", "endpoint_id": "k8s_observer/e6d880c9-13a0-45e7-a3a1-27fa021802b4/webserver(80)", "config": {"collection_interval":"20s","endpoint":"http://`endpoint`/nginx_status"}}
....
2024-10-04T13:53:02.550Z	info	receivercreator@v0.110.0/observerhandler.go:115	starting receiver	{"kind": "receiver", "name": "receiver_creator/metrics", "data_type": "metrics", "name": "redis/e6d880c9-13a0-45e7-a3a1-27fa021802b4_6379", "endpoint": "10.244.0.52:6379", "endpoint_id": "k8s_observer/e6d880c9-13a0-45e7-a3a1-27fa021802b4/redis(6379)", "config": {"collection_interval":"20s","endpoint":"10.244.0.52:6379"}}
2024-10-04T13:53:03.646Z	info	MetricsExporter	{"kind": "exporter", "data_type": "metrics", "name": "debug", "resource metrics": 2, "metrics": 30, "data points": 38}
2024-10-04T13:53:06.547Z	debug	memorylimiter/memorylimiter.go:181	Currently used memory.	{"kind": "processor", "name": "memory_limiter", "pipeline": "traces", "cur_mem_mib": 39}
2024-10-04T13:53:11.547Z	debug	memorylimiter/memorylimiter.go:181	Currently used memory.	{"kind": "processor", "name": "memory_limiter", "pipeline": "traces", "cur_mem_mib": 39}
2024-10-04T13:53:16.546Z	debug	memorylimiter/memorylimiter.go:181	Currently used memory.	{"kind": "processor", "name": "memory_limiter", "pipeline": "traces", "cur_mem_mib": 40}
```


**More testing target Pods**
[busybox-multi-logs.txt](https://github.com/user-attachments/files/17523701/busybox-multi-logs.txt)
[busybox-single-logs.txt](https://github.com/user-attachments/files/17523702/busybox-single-logs.txt)
[redis-nginx-mutli.txt](https://github.com/user-attachments/files/17523703/redis-nginx-mutli.txt)
[redis-single.txt](https://github.com/user-attachments/files/17523704/redis-single.txt)
[redis-single-logs.txt](https://github.com/user-attachments/files/17523705/redis-single-logs.txt)

ChrsMark committed:
[receiver/receiver_creator] Add support for metrics' hints

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

ChrsMark commented:
@dmitryax I have opened this for review, please feel free to take a look when you get the chance.

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,223 @@
+		return nil, nil
+	}
+
+	if builder.config.Enabled {
```
Is this guard needed? I think `createScraperTemplateFromHints` will be only called if this is true: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-b4ba0783a3cc06542046d42bf63024247eba48e06e186b87dad0b0a1ed189e95R92

ChrsMark replied:
👍🏽 leftover from the refactoring. Thank's!

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,223 @@
+	userConfMap["signals"] = signals
+
+	if p, ok := env["port"]; ok {
+		port = p.(uint16)
+		if port == 0 {
+			return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+		}
```
Wdyt if we wrap the interface casting to prevent any possible panic?
```suggestion
		port, ok = p.(uint16)
		if !ok || port == 0 {
			return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
		}
```

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,223 @@
+	} else {
+		return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+	}
+	subreceiver, err := newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, podUID, port), userConfMap)
+	if err != nil {
+		builder.logger.Error("error adding subreceiver", zap.Any("err", err))
+		return nil, err
+	}
+
+	builder.logger.Debug("adding hinted receiver", zap.Any("subreceiver", subreceiver))
+	return &subreceiver, nil
```
The following log messages are already covered by the caller function, maybe we can simplify it with:

```suggestion
	return newReceiverTemplate(fmt.Sprintf("%v/%v_%v", subreceiverKey, podUID, port), userConfMap)
```
(newReceiverTemplate needs to be changed to return a reference)

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,223 @@
+
+	signalsAnn := getHintAnnotation(annotations, signalsHint, portName)
+	signals := getSignalsConf(signalsAnn)
+	userConfMap["signals"] = signals
```
Wdyt if we extend the userConfMap to a struct with the unparsed raw configuration of the receiver `map[string]any` and the actual known fields like signal?

I think this would prevent some type casts and deleting unused fields: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-b4ba0783a3cc06542046d42bf63024247eba48e06e186b87dad0b0a1ed189e95R102

ChrsMark replied:
Actually, I made `signals` a field of the [`receiverTemplate`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0898be1e204e10dc8127699743f6bf9a62a529ed/receiver/receivercreator/config.go#L55) since it has similar "weight" with the `rule` field. In the future we could even expose this as a setting to the users so as it can be applied to static receiver_creator's configs. Let me know what you think.

rogercoll replied:
That looks good, thanks :+1:

rogercoll commented on receiver/receivercreator/README.md:
```diff
@@ -438,3 +438,186 @@ service:
+opentelemetry.io.discovery/config.xyz: 'abc'
+
+# nested configs
+opentelemetry.io/discovery/config.nested: |
+  zyz: foo
+  bar: baz
```
To prevent reserving the `nested` keyword for components configuration, could we just use the config keyword instead:
```suggestion
opentelemetry.io/discovery/config: |
  zyz: foo
  bar: baz
```

ChrsMark replied:
Hmm `.nested` here is not a fixed "hint"/"indicator" it's just a random example config's key. I can change it to sth else to avoid confusion :).

ChrsMark replied:
Fixed as per discussions at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34427.

rogercoll commented on receiver/receivercreator/README.md:
```diff
@@ -438,3 +438,186 @@ service:
+
+#### Enable/disable discovery
+
+`opentelemetry.io.discovery/enabled` (example: `"true"`)
+
+By default `"true"`.
```
I think this can be a bit misleading with the current logic's; what will happen if a user edits an annotation from `true` to `false` after the receiver creator has already created the hints' receiver? I would expect the receiver_creator to shut down the created receiver. 

Likewise, I am not sure if the receiver_creator is able to perform the latest, what about if for the moment we remove `opentelemetry.io.discovery/enabled` and assume that if the other fields are defined the user expects the receiver to be created?

ChrsMark replied:
> I would expect the receiver_creator to shut down the created receiver

I believe that's exactly what will happen. If Pod's annotations change then [`OnChange`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fa20d6a9e8c34be8fef2de806a2a7350b3c62c32/receiver/receivercreator/observerhandler.go#L150) is called  which stops+starts the receiver(s). However if the annotation `opentelemetry.io.discovery/enabled: "false"` is added the receivers won't start again because `start` will take through the whole annotations' evaluation again and will take into account the newly added annotation.

rogercoll replied:
My assumptions were wrong then, thanks for the detailed explanation and testing it. LGTM

rogercoll reviewed (commented):
Great work @ChrsMark! :rocket: Left some comments/doubts

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,250 @@
+				return nil, nil
+			}
+		}
+		podUID = endpointPod["uid"].(string)
+		podName = endpointPod["name"].(string)
+		podNs = endpointPod["namespace"].(string)
```
Will these fields be always present? Wdyt adding some type-assertions or comment explaining why this fields will be always string and present in the k8s api?

ChrsMark replied:
Good point. Actually I realize we can just use the observer's structs directly: 216998e34f95f0a8a4a975c2f4ab9290176d2222. 

However, that might require the changes in the `PodContainer` to happen in a separate PR 🤔 . (change from `container_name` to `name` etc)

rogercoll commented on receiver/receivercreator/README.md:
```diff
@@ -439,3 +439,296 @@ service:
+
+#### Define scraper
+
+`io.opentelemetry.discovery.metrics/scraper` (example: `"nginx"`)
```
Are scrapers names with ID supported? E.g. `nginx/server`?

Although it would be great to support this, we can keep it for future improvements.

ChrsMark replied:
We could figure out how to support this in the future, yeap.

At the moment the ID is constructed combining the provided name, the Pod's UID and the port. This would ensure the uniqueness of the ID for the Collector instance I think.

Do you have any specific example of how `nginx/server` would be used by a user?

rogercoll commented on receiver/receivercreator/README.md:
````diff
@@ -439,3 +439,296 @@ service:
+
+```yaml
+receiver_creator/logs:
+  watch_observers: [ k8s_observer ]
+  hints:
+    k8s:
````
Is the `k8s` configuration key needed? Somehow, I feel like we are creating a configuration relationship between the `watch_observers` (interface) and the `hints` configuration.

Would it make sense to document that `hints` is only supported for k8s at the moment? If in the future can we have other "hints/annotations" observers, they could reuse the same "hints" config (or not):

```yaml
// reusing hints config across observers
receiver_creator/logs:
  watch_observers: [ k8s_observer, file_observer ]
  hints:
    metrics:
      enabled: true

// different configs
receiver_creator/logs/k8s:
  watch_observers: [ k8s_observer ]
  hints:
    metrics:
      enabled: true
receiver_creator/logs/file:
  watch_observers: [ file_observer ]
  hints:
    metrics:
      enabled: true
```

ChrsMark replied:
The idea is that we can configure the hints feature for multiple environments (in the future) like docker, k8s etc in a way that the configuration can be specific per environment.

So we could have sth like the following:

```yaml
// different configs
receiver_creator/metrics:
  watch_observers: [ k8s_observer, docker_observer ]
  hints:
    k8s: 
      metrics:
        enabled: true
    docker:
      metrics:
        # docker specific settings
```

Happy to reconsider this though if it doesn't make sense.

> Would it make sense to document that hints is only supported for k8s at the moment?

It is mentioned at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-4127365c4062a7510fb7fede0fa239e9232549732898303d94c12fef0433d39dR448

dmitryax replied:
I agree that the nested configuration seems confusing. Users can have separate `receiver_creator` instances with different `watch_observers` if they want to split the behavior between docker/k8s/etc.

I'm also not sure about using the _hints_ term everywhere. It seems not very descriptive to me. What about using `discovery` term instead, given that this is something we already use in the annotations?

Maybe something like this:
```yaml
receiver_creator:
  watch_observers: [k8s_observer]
  discovery:
    enabled: true
    ignore_receivers: [filelog]
```

This would mean only allow creation of metrics related receivers from the pod annotations. Also, if a user finds any issues with let's say `nginx` metrics, they can easily add it to `ignore_receivers`...

ChrsMark replied:
Cool, it seems we can unify then. I also like the idea of `ignore_receivers` since it provides the option to Collector admins  to disable specific receivers for security reasons. I will update the PR accordingly.

rogercoll reviewed (approved):
LGTM, just some nits

ChrsMark commented on receiver/receivercreator/README.md:
```diff
@@ -439,3 +439,311 @@ service:
+The hints are evaluated per container by extracting the annotations from each [`Port` endpoint](#Port) that is emitted. 
+
+
+### Supported logs annotations
+
+This feature enables `filelog` receiver along with the `container` parser in order to collect logs from the discovered
+Pods.
```
Splitting is fine, I just thought that putting the whole feature together would help the review process but I can only keep metrics' part for now 👍🏽 . I will update the PR.

ChrsMark commented on extension/observer/endpoints.go:
```diff
@@ -223,11 +223,11 @@ func (p *Pod) Type() EndpointType {
 type PodContainer struct {
 	// Name of the container
-	Name string
+	Name string `mapstructure:"container_name"`
```
I don't think that's a breaking change. That's exactly how the the `PodContainer` is serialized by the `Env()` method right bellow:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/806a4ef48cde1dd296dabf2c117bc6df4025c409/extension/observer/endpoints.go#L235-L242

So I think that's compatible with what we expose to the users today. Or do I miss something here?

dmitryax replied:
Oh, I see. Can you please point where it's being used for deserialization now?

ChrsMark replied:
In the following 2 places of the introduced code:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-2db19e270904db9b7d87966c13aaa399b48566a81f520f91fa8af8032fdc9827R95

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/files#diff-2db19e270904db9b7d87966c13aaa399b48566a81f520f91fa8af8032fdc9827R75

ChrsMark commented on receiver/receivercreator/README.md:
```diff
@@ -439,3 +439,311 @@ service:
+       enabled: true
+       # collect logs from all containers unless discovery is explicitly disabled
+       # for a container using the discovery hint
+       default: true
```
Let's deal with that in the follow-up PR that will add logs support.

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		if pod.Annotations != nil {
+			if len(pod.Annotations) == 0 {
+				return nil, nil
+			}
+		}
```
```suggestion
		if len(pod.Annotations) == 0 {
			return nil, nil
		}
```
"Slice, or map: the number of elements in v; if v is nil, len(v) is zero." reference - https://pkg.go.dev/builtin#len

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,178 @@
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, "endpoint")
```
```suggestion
	defaultEndpoint := getStringEnv(env, endpointConfigKey)
```

Wdyt of using the global variable instead?

ChrsMark replied:
Good point, actually we can re-use the one from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3ca96862b9205e447cf56c855505c714202cca17/receiver/receivercreator/config.go#L20.

rogercoll commented on receiver/receivercreator/hints.go:
```diff
@@ -0,0 +1,178 @@
+	logger *zap.Logger) userConfigMap {
+	conf := userConfigMap{}
+	if defaultEndpoint != "" {
+		conf["endpoint"] = defaultEndpoint
```
```suggestion
		conf[endpointConfigKey] = defaultEndpoint
```

rogercoll reviewed (approved):
LGTM, should we rename the "hints.go" file into "discovery.go"? (to be aligned with the config option)

ChrsMark commented:
Thank's @rogercoll. I have applied your suggestions and also rebased the branch to resolve the conflicts. Also squashed the commits into a single one to better manage (future) conflicts.

/cc @dmitryax

ChrsMark commented:
@dmitryax as per our discussions [cc869c6](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/commits/cc869c6b3a862fda4e08016f51e27a43a4caa9b0) ensures that the provided configuration from the annotation only targets the discovered Pod.

Examples of valid configs:

1. 
```yaml
endpoint: "http://`endpoint`/nginx_status"
```

2. 
```yaml
endpoint: "`endpoint`"
```

Example of not valid config:

```yaml
endpoint: "https://8.8.8.8:8080" # endpoint that does not match the target Pod 
```

Feel free to have a look.

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,186 @@
+
+func discoveryMetricsEnabled(annotations map[string]string, hintBase string, scopeSuffix string) bool {
+	hintVal := getHintAnnotation(annotations, hintBase, discoveryEnabledHint, scopeSuffix)
+	return hintVal == "true" || hintVal == ""
```
This seems like undocumented support of `""`. I think we should avoid that. Let's support only explicit `true` values?

dmitryax replied:
Hm.. Looks like `getHintAnnotation` actually returns `""` is the annotation isn't found. So `hintVal == ""` makes it a bug AFAIU

ChrsMark replied:
The idea here is that if the `io.opentelemetry.discovery.metrics/enabled` is not explicitly disabled we default to `true`. This is to reduce the number of annotations users need to specify. Only defining the scraper annotation should be enough without explicitly adding the `io.opentelemetry.discovery.metrics/enabled: true`.

This is mentioned at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/451da6ca3a41a363af963e9d2b3a2087c4fbfac5/receiver/receivercreator/README.md?plain=1#L469.

dmitryax replied:
Doesn't `io.opentelemetry.discovery.metrics/enabled: ` give `nil` so users need to specify `""` values explicitly as `io.opentelemetry.discovery.metrics/enabled: ""`. I disagree that this provides more benefit than confusion. Users need to specify the annotation, which is typically copy/paste. So it's better to be explicit on what needs to be copy/pasted. `getHintAnnotation` returning `""` still seems like a bug if we accept `""` here.

ChrsMark replied:
All right, there is a case to be made here for having the `io.opentelemetry.discovery.metrics/enabled` as required annotation hint. Thinking of it again I agree it would be more consistent and would also be future proof if and when the new `Discovery()` interface of scrapers will happen.

I'm gonna change it in this way.

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,186 @@
+// Scraper configurations are only created for Port Endpoints.
+// TODO: Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *k8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
```
nit: this declaration seems redundant

ChrsMark replied:
done: [046754a](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617/commits/046754a81ec3f31fa2a34a3725b796a4a5aa7804)

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,186 @@
+		if err != nil {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		if len(pod.Annotations) == 0 {
+			return nil, nil
+		}
```
This must be redundant if we had `getHintAnnotation` handled properly

ChrsMark replied:
I've made the `getHintAnnotation` to also return if hint was found or not explicitly.
I've also removed that `len(pod.Annotations) == 0` check.

ChrsMark committed:
make enablement hint mandatory

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

ChrsMark commented:
@dmitryax based on the discussion at https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35617#discussion_r1845250244 I have changed the `io.opentelemetry.discovery.metrics/enabled` hint to mandatory. 

PTAL.

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,185 @@
+	}
+	return k8sHintsBuilder{
+		logger:          logger,
+		config:          config,
```
Why do we need to store the full config if we have a separate field for `ignoreReceivers`? Can be just enabled right?

ChrsMark replied:
Must be a leftover after some of the re-designs, good catch! Remove that now, thank's!

dmitryax reviewed (approved):
LGTM with a nit

ChrsMark committed:
remove unused config field

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

ChrsMark commented:
Thank's @dmitryax, I've removed the redundant config field. It should be good to go now.

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,183 @@
+)
+
+// hintsTemplatesBuilder creates configuration templates from provided hints.
+type hintsTemplatesBuilder interface {
+	createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error)
+}
```
This interface doesn't seem like being used anywhere, right?

ChrsMark replied:
right now it's only implemented by the `k8sHintsBuilder`. The idea is to give the option for implementing it in the various different environments like k8s, docker etc. This feature would make sense for docker as well.

dmitryax replied:
But `k8sHintsBuilder` is not being used at all. Existing functions operate on the struct instead. I don't see why we need to keep it around. It can be added when needed

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,183 @@
+		return userConfigMap{}
+	}
+
+	if !strings.Contains(confEndpoint, defaultEndpoint) {
```
I have a couple of questions here:

1. Will this work for annotation with "`endpoint`" string in it? like the one we have in the readme
```
io.opentelemetry.discovery.metrics/config: |
  endpoint: "http://`endpoint`/nginx_status"
```

2. I think "contains" isn't a strong enough check. Someone can configure the annotation to be 
"http://another-endpoint/nginx_status?my-endpoint:8888" and if collector will scrape "another-endpoint". I think we should use URI pattern check here instead

ChrsMark replied:
Right to both. The validation had to be improved indeed. Fixed that in the last commit.

ChrsMark committed:
improve endpoint validation

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

ChrsMark committed:
gofmt

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

dmitryax commented on receiver/receivercreator/discovery.go:
```diff
@@ -0,0 +1,220 @@
+		}
+		uri = u
+	}
+	// target endpoint can come in form ip:port. In that case we fix the uri
+	// temporarily with adding http scheme
+	if uri.Scheme == "" {
+		u, err := url.Parse("http://" + endpoint)
+		if err != nil {
+			return fmt.Errorf("could not parse enpoint")
+		}
+		uri = u
+	}
```
This block seems redundant

ChrsMark replied:
Removed.

ChrsMark committed:
remove unused code

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

ChrsMark committed:
Merge branch 'main' into hints
